### ISSUE TITLE
TURN protocol leaks username

### ISSUE THREAD

**Issue by MilanKral** [cmeng-git/atalk-android#144]
Current mobile networks with NAT many times require use of TURN servers.
Most TURN server authenticate users using username and password.

Because aTalk only supports UDP protocol for TURN it leaks the username - username is transferred in clear text.  When "Auto discover STUN/TURN" is used to gather TURN servers, then the TURN username is the Jabber username, so Jabber username is leaked.

Please consider enabling support for TLS, DTLS for TURN.

Please consider implementing support for XEP-0215, which allows the Jabber server to publish short-term TURN login username and password.

**Comment by MilanKral** [comment]
Here is a proposal for TLS, DTLS support.
https://github.com/MilanKral/atalk-android/tree/TURN_TLS

But I still can't get ICE4J to use TLS, DTLS.

**Comment by cmeng-git** [comment]
Thanks for your contribution. I have manually pulled in all the source changes for the next aTalk release.

**Comment by ericschdt** [comment]
> Thanks for your contribution. I have manually pulled in all the source changes for the next aTalk release.

What's left to make DTLS working? In 2.2.4 the option is grayed out and not selectable.

**Comment by cmeng-git** [comment]
I have continue your question in its original thread.

**Comment by cmeng-git** [comment]
> But I still can't get ICE4J to use TLS, DTLS.

Just  checked in the source with ice4j-2.0.0-20181024.160538-12 upgraded to the latest ice4j-2.0.0-20190607.184546-36. No sure if this helps in TLS and DTLS

**Comment by MAGDIA** [comment]
I need NAT config for Atalk

MDS

**Comment by MAGDIA** [comment]
Hello
Sir

I referred to the online help for configuring NAT between two different mobile networks (example between the Orange mobile network and Teletel), but without success.

Let me explain :
1 - When two smartphones on two different mobile networks (example between the Orange mobile network and Teletel mobile) are called, communication is established but there is no sound (ZRTP does not turn green) so the correspondents do not get along.

2 - When two smartphones on the same mobile network (eg the Orange mobile network alone) call each other, communication is established and there is sound (ZRTP turns green) so the correspondents get along very well.

To fix the NAT I have tried the STUN Google = stun.l.google.com: 19302 and our own STUN = server2magdia.africa.com:3478 without success.

Always to correct the NAT I also tried some Atalk versions (versions: 1001, 1053, 1062, 1.6.6, 2.3.4, 2.4.1, and the latest = 2.4.3}, but also without success.

In short, the only problem I have with Atalk, this product that I love so much, is correcting the NAT.

I sincerely ask you to help me solve this problem.

If you need two clients (openfire users) to test yourself, let me know.

The version of my Openfire server is 4.2.3.

I also have a question:
Which XMPP server does Atalk not have a NAT problem with (i.e. on two different mobile networks, the correspondents get along very well) ?.

 Thank you for your prompt reaction

Regards,

MAGDIA

**Comment by cmeng-git** [comment]
All the network NAT settings are defined in ICE:
[STUN, TURN, ICE & Jingle for NAT Traversal in media call setup](https://cmeng-git.github.io/atalk/faq.html#media_06)

I too encountered problem when making call between two devices on two different networks.  In my case, ice4j failed to establish connection via all the given transport candidates,
```
2020-11-24 18:21:23.767 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: [fe80::8961:5b15:60cd:e093]:5000/udp/host -> [fe80::f025:b7ff:fe9c:3c5]:5000/udp/host (audio.RTP), failing.
2020-11-24 18:21:23.792 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: [fe80::8961:5b15:60cd:e093]:5000/udp/host -> [fe80::f225:b7ff:fe9c:3c5]:5000/udp/host (audio.RTP), failing.
2020-11-24 18:21:23.815 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.135.253.160:5000/udp/host -> 192.168.1.30:5000/udp/host (audio.RTP), failing.
2020-11-24 18:21:23.839 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.161.160.121:5000/udp/host -> 192.168.1.30:5000/udp/host (audio.RTP), failing.
2020-11-24 18:21:23.860 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.135.253.160:5000/udp/host -> 42.60.7.13:5000/udp/srflx (audio.RTP), failing.
2020-11-24 18:21:23.888 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.161.160.121:5000/udp/host -> 42.60.7.13:5000/udp/srflx (audio.RTP), failing.
2020-11-24 18:21:23.891 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() CheckList will failed in a few seconds if no succeeded checks come
2020-11-24 18:21:24.670 16746-22251/org.atalk.android D/AudioTrack: stop(3649): called with 22050 frames delivered
2020-11-24 18:21:25.039 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: [fe80::8961:5b15:60cd:e093]:5001/udp/host -> [fe80::f025:b7ff:fe9c:3c5]:5001/udp/host (video.RTP), failing.
2020-11-24 18:21:25.060 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: [fe80::8961:5b15:60cd:e093]:5001/udp/host -> [fe80::f225:b7ff:fe9c:3c5]:5001/udp/host (video.RTP), failing.
2020-11-24 18:21:25.088 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.135.253.160:5001/udp/host -> 192.168.1.30:5001/udp/host (video.RTP), failing.
2020-11-24 18:21:25.110 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.161.160.121:5001/udp/host -> 192.168.1.30:5001/udp/host (video.RTP), failing.
2020-11-24 18:21:25.136 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.135.253.160:5001/udp/host -> 42.60.7.13:5001/udp/srflx (video.RTP), failing.
2020-11-24 18:21:25.163 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() timeout for pair: 10.161.160.121:5001/udp/host -> 42.60.7.13:5001/udp/srflx (video.RTP), failing.
2020-11-24 18:21:25.165 16746-22288/org.atalk.android I/aTalk: [47207] org.ice4j.ice.ConnectivityCheckClient.log() CheckList will failed in a few seconds if no succeeded checks come
```

From aTalk history log, I saw that aTalk v2.2.5 is released with an upgrade to ice4j library i.e. 
Port ice4j-2.0.0-20181024.160538-12 to ice4j-2.0.0-20190607.184546-36
I am not sure if this is the cause of the problem, or the NAT firewall is blocking all the traffic. Need more time to investigate.

Meantime to overcome the problem, you may have to setup a Relay on your server to take care of the problem.

**Comment by cmeng-git** [comment]
aTalk v3.0.3 has implemented the supported for XEP-0215.

`Please consider implementing support for XEP-0215, which allows the Jabber server to publish short-term TURN login username and password.`
